Don't calculate hashes for schema versions in Raft mode

kbr-scylla opened the pull request:
When performing a schema change through group 0, extend the schema mutations with a version that's persisted and then used by the nodes in the cluster in place of the old schema digest, which becomes horribly slow as we perform more and more schema changes (#7620).

If the change is a table create or alter, also extend the mutations with a version for this table to be used for `schema::version()`s instead of having each node calculate a hash which is susceptible to bugs (#13957).

When performing a schema change in Raft RECOVERY mode we also extend schema mutations which forces nodes to revert to the old way of calculating schema versions when necessary.

We can only introduce these extensions if all of the cluster understands them, so protect this code by a new cluster/schema feature, `GROUP0_SCHEMA_VERSIONING`.

Fixes: #7620
Fixes: #13957

---

This is a reincarnation of PR scylladb/scylladb#15331. The previous PR was reverted due to a bug it unmasked; the bug has now been fixed (scylladb/scylladb#16139). Some refactors from the previous PR were already merged separately, so this one is a bit smaller.

I have checked with @Lorak-mmk's reproducer (https://github.com/Lorak-mmk/udt_schema_change_reproducer -- many thanks for it!) that the originally exposed bug is no longer reproducing on this PR, and that it can still be reproduced if I revert the aforementioned fix on top of this PR.

avikivity commented on db/schema_tables.cc:
```diff
@@ -354,6 +355,19 @@ schema_ptr scylla_tables(schema_features features) {
+            // before the GROUP0_SCHEMA_VERSIONING feature was enabled (either inside or outside group 0).
+            // In this case, for non-system tables, `version` is null and `schema::version()` will be a hash.
+            sb.with_column("committed_by_group0", boolean_type);
+            offset |= 0b100;
```
Please reserve 0b100 for scylla-enterprise. It was wrong of us to pick a number from the same range, but we did.

avikivity replied:
But, I think that after this series, we care a lot less about digest hashing. The next time we update the schema's schema, the update will be propagated via raft rather than each node calculating a new digest and hoping it will match.

kbr-scylla replied:
This is a local table, its schema is not propagated (by Raft or otherwise). Its schema version is calculated below: `system_keyspace::generate_schema_version(id, offset)`.

I can reserve `0b100`, although I think that any modifications to this table should always go through open source first for simplicity.

avikivity commented:
Looks good.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:white_check_mark: - Unit Tests
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4993/
- Duration: 3 hr 2 min
- Builder: i-096b62f88c203d6ac

kbr-scylla commented on db/system_keyspace.cc:
```diff
@@ -2406,6 +2406,23 @@ future<mutation> system_keyspace::get_group0_history(distributed<replica::databa
+future<mutation> system_keyspace::get_group0_schema_version() {
+    auto s = _db.find_schema(db::system_keyspace::NAME, db::system_keyspace::SCYLLA_LOCAL);
+
+    partition_key pk = partition_key::from_singular(*s, "group0_schema_version");
```
scylla_local has many partitions. It's basically a string->string key-value store.

kbr-scylla commented:
v2:
- added group 0 schema versioning section to raft docs
- reserved `0b100` in `schema_ptr scylla_tables(schema_features features)` for enterprise

Diff: https://github.com/scylladb/scylladb/compare/c3ebead9fda7eb01bafffbc4e8d8bba344d9a31d..242d85e5c606965c1b7fd44c48837cead94911dc

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5004/)
:white_check_mark: - [Sanity Tests](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6362/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5004/)
#### Build Details:
- Duration: 2 hr 11 min
- Builder: spider8.cloudius-systems.com

kbr-scylla commented:
v3:
- rebased
- in the spirit of https://github.com/scylladb/scylladb/pull/16192, I replaced parts of the tests which did non-Raft -> Raft upgrade, with going through RECOVERY instead, so I don't interrupt @patjed41's work on issue https://github.com/scylladb/scylladb/issues/15854
- while running the tests, I observed that after a schema change inside RECOVERY mode, sometimes driver's schema agreement wait would timeout (but the test would pass, just run longer). After investigation, it turned out that the tombstone that I'm inserting into `system.scylla_local` under `group0_schema_version` key (as described in the "migration_manager: store group0_schema_version in scylla_local during schema changes" commit) is getting lost in schema pulls, because it's being GCd before the pull happens (ref: https://github.com/scylladb/scylladb/issues/15607). I improved the test to catch this problem (i.e. fail when it happens), and it this by preventing the tombstone from being GCd with the following diff:
```diff
-        : atomic_cell::make_dead(guard.write_timestamp(), gc_clock::now());
+        // Use gc_clock::time_point::max() for deletion time so the tombstone is not GCd
+        // (so that snapshot transfers / schema pulls observe it)
+        : atomic_cell::make_dead(guard.write_timestamp(), gc_clock::time_point::max());
```

avikivity commented:
Will this change of deletion_time cause tombstones to be accumulated?

```sh
while true; do
   tab=tab_$RANDOM_$RANDOM
   cqlsh -e "CREATE TABLE $tab ..."
   cqlsh -e "DROP TABLE $tab"
done
```

Creates an infinite number of tombstones which are never garbage collected.

Is it legal to just change deletion_time like that? Seems like using now() is the right thing.

kbr-scylla commented:
> Will this change of deletion_time cause tombstones to be accumulated?

It's at most one tombstone -- under the `group0_schema_version` key in `system.scylla_local`. And it's overwritten by a live cell as soon as we perform schema change in Raft mode (outside recovery).

> Is it legal to just change deletion_time like that? Seems like using now() is the right thing.

Why not? This time is used by nodes to decide when it is safe to GC the tombstone. They compare the deletion_time with their now() + table's GC duration. By setting deletion_time to max() I simply prevent the tombstone from being GCd ever.

kbr-scylla commented:
I considered the alternative of storing a live cell with a special value (like UUID=0) instead of such non-GCable tombstone, but the tombstone solution seems more elegant.

kbr-scylla commented:
> They compare the deletion_time with their now() - table's GC duration

Let me check how this comparison is done just in case... if it's something like `deletion_time + GC duration <= now()`, we might have overflow problem due to using `deletion_time == max()`. But if there was overflow and `deletion_time + GC duration` wrapped, the tombstone would be GCd, which the test would've catched -- but it didn't. So it's probably done like `deletion_time <= now() - GC duration`.

avikivity commented:
Ok it's reasonable. But @tgrabiec needs to review too.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5061/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5061/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6423/)
#### Build Details:
- Duration: 2 hr 22 min
- Builder: spider7.cloudius-systems.com

kbr-scylla commented:
ping, @tgrabiec please review

kbr-scylla commented:
v4:
- instead of using `gc_clock::time_point::max()` for the tombstone in `system.scylla_local` (`group0_schema_version` key), use `gc_clock::now()` like everyone else, but increase `gc_grace` for the table from 0 to 7 days, like for schema tables.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5136/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6515/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5136/)
#### Build Details:
- Duration: 2 hr 13 min
- Builder: spider2.cloudius-systems.com

kbr-scylla commented:
@avikivity / @tgrabiec ping, please review/merge
